In [1]:
import pandas as pd
import psycopg2
from psycopg.errors import ForeignKeyViolation
import os
import tabulate
import numpy as np

## 1. Creating Dataframes for phyla, classes, subclasses, orders, families, genera, species and specie-synonyms from the WFO csv

In [3]:
columns_to_keep = ['taxonID',
                   'scientificName',
                   'taxonRank',
                   'parentNameUsageID',
                   'family',
                   'genus',
                   'nomenclaturalStatus',
                   'taxonomicStatus',
                   'acceptedNameUsageID'
                   ]

# WFO Database saved as csv file
input_csv = 'classification.csv'

# transforms taxonID to an integer ID (e.g wfo-0001302018 --> 1302018)
def taxon_number(taxon_id):
    if taxon_id[:3] == 'wfo':
        return int(taxon_id[4:])
    else:
        return 0

In [52]:
# list of english phyla names corresponding to the scientific names
phyla_english_names = [('Lycopodiophyta', 'Lycopods'),
                       ('Marchantiophyta', 'Liverworts'),
                       ('Angiosperms', 'Flowering plants'),
                       ('Anthocerotophyta', 'Hornworts'),
                       ('Bryophyta', 'Mosses'),
                       ('Polypodiophyta', 'Ferns'),
                       ('Pinophyta', 'Conifers and allies'),
                       ('Ginkgophyta', 'Ginkgo'),
                       ('Cycadophyta', 'Cycadophyta')
                       ]
phyla_english_names_df = pd.DataFrame(phyla_english_names, columns=['scientificName', 'phylum_englishName'])

In [5]:
wfo_df = pd.read_csv(input_csv,
                     sep='\t',
                     encoding='ANSI',
                     usecols=columns_to_keep,
                     converters={0: taxon_number, 5: taxon_number, 19: taxon_number}
                     )

### Possible values for nomenclaturalStatus and taxonomicStatus:

In [6]:
display(wfo_df['nomenclaturalStatus'].value_counts())
display(wfo_df['taxonomicStatus'].value_counts())
# print(wfo_df['taxonRank'].value_counts())

nomenclaturalStatus
Valid           840842
Illegitimate     28364
Invalid          18175
Superfluous       1561
Conserved          521
Rejected           432
Name: count, dtype: int64

taxonomicStatus
Synonym      943255
Accepted     444907
Unchecked    187975
Name: count, dtype: int64

#### We will keep only Conserved and Valid nomenclaturalStatus, and Accepted taxonomicStatus. But first we save all the species which are synonyms of accepted species in a separate dataframe:

In [762]:
# species_synonyms_df = wfo_df.loc[(wfo_df['nomenclaturalStatus'] == 'Valid') & (wfo_df['taxonomicStatus'] == 'Synonym') & (wfo_df['taxonRank'] == 'species'), ['taxonID', 'scientificName', 'acceptedNameUsageID']]
# display(species_synonyms_df)

taxonID       scientificName  acceptedNameUsageID
15          1302026   Kobresia bucharica               345171
27          1302039  Carex graciliformis               344146
30          1302042  Carex kattegatensis               352303
35          1302048     Carex musartiana               350145
74          1302089  Pycreus nilagiricus               373599
...             ...                  ...                  ...
1575565  1000057514      Erica gordoniae               673433
1575566  1000057515       Erica eweriana               671867
1575995  1000057833       Azalea punicea              1229780
1576093  1000057862  Pilostyles pringlei              1129097
1576128  1200069471     Bronnia diguetii               691655

[248004 rows x 3 columns]

In [7]:
wfo_df = wfo_df[(wfo_df['nomenclaturalStatus'] == 'Valid') | (wfo_df['nomenclaturalStatus'] == 'Conserved')]
wfo_df = wfo_df[wfo_df['taxonomicStatus'] == 'Accepted']
wfo_df = wfo_df.sort_values('scientificName')
display(wfo_df)

taxonID     scientificName taxonRank  parentNameUsageID  \
1405743  4000000001                 Aa     genus         7000000429   
316954       319089      Aa achalensis   species         4000000001   
756946       760991     Aa argyrolepis   species         4000000001   
918422       922666      Aa aurantiaca   species         4000000001   
923805       928062        Aa calceata   species         4000000001   
...             ...                ...       ...                ...   
1442733  4000046085  × Zygosepescalum     genus         7000000429   
1441514  4000044858      × Zygostylis     genus         7000000429   
1443289  4000046644       × Zygotoria     genus         7000000429   
1443330  4000046685      × Zygowarrea     genus         7000000429   
1442586  4000045938       × Zygozella     genus         7000000429   

              family           genus nomenclaturalStatus taxonomicStatus  \
1405743  Orchidaceae              Aa               Valid        Accepted   
316954   Orchidaceae              Aa               Valid        Accepted   
756946   Orchidaceae              Aa               Valid        Accepted   
918422   Orchidaceae              Aa               Valid        Accepted   
923805   Orchidaceae              Aa               Valid        Accepted   
...              ...             ...                 ...             ...   
1442733  Orchidaceae  Zygosepescalum               Valid        Accepted   
1441514  Orchidaceae      Zygostylis               Valid        Accepted   
1443289  Orchidaceae       Zygotoria               Valid        Accepted   
1443330  Orchidaceae      Zygowarrea               Valid        Accepted   
1442586  Orchidaceae       Zygozella               Valid        Accepted   

         acceptedNameUsageID  
1405743                    0  
316954                     0  
756946                     0  
918422                     0  
923805                     0  
...                      ...  
1442733                    0  
1441514                    0  
1443289                    0  
1443330                    0  
1442586                    0  

[444893 rows x 9 columns]

In [8]:
print(wfo_df['nomenclaturalStatus'].value_counts())
print(wfo_df['family'].value_counts())
print(wfo_df['genus'].value_counts())


nomenclaturalStatus
Valid        444500
Conserved       393
Name: count, dtype: int64
family
Asteraceae          42662
Orchidaceae         35644
Fabaceae            30661
Rubiaceae           16761
Poaceae             14120
                    ...  
Culcitaceae             3
Saelaniaceae            3
Agdestidaceae           3
Tetracarpaeaceae        3
Chenopodiaceae          1
Name: count, Length: 717, dtype: int64
genus
Hieracium          4465
Astragalus         3741
Carex              2826
Euphorbia          2562
Begonia            2522
                   ... 
Zygomatophyllum       1
Zygolum               1
Zygogardmannia        1
Zygodisanthus         1
Zygocella             1
Name: count, Length: 17874, dtype: int64


Now that we reduced the dataframe to valid or conserved entries, we list the possible taxonRank in taxonomic order:
<br/><br/>

|             | Any   | Valid  | Conserved|
|-------------|-------|--------| ----------- |
| kingdom     | 1     | 1      | 0 |
| subkingdom  | 2     | 2      | 0 |
| phylum      | 9     | 9      | 0 |
| class       | 16    | 16     | 0 |
| subclass    | 26    | 26     | 0 |
| superorder  | 2     | 2      | 0 |
| order       | 142   | 141    | 1 |
| family      | 717   | 698    | 19 |
| subfamily   | 90    | 89     | 1 |
| tribe       | 84    | 84     | 0 |
| subtribe    | 170   | 170    | 0 |
| genus       | 17910 | 17598  | 312 |
| subgenus    | 71    | 71     | 0 |
| section     | 267   | 267    | 0 |
| subsection  | 86    | 86     | 0 |
| series      | 22    | 22     | 0 |
| subseries   | 2     | 2      | 0 |
| species     | 377223 | 377164 | 59 |
| prole       | 1     |        |   |
| subspecies  | 23371 |        |   |
| variety     | 23821 |        |   |
| subvariety  | 3     |        |   |
| form        | 857   |        |   |

In [10]:
wfo_df_grouped_by_taxonRank = wfo_df.groupby('taxonRank')

In [13]:
parent_df = wfo_df.filter(['taxonID', 'scientificName', 'taxonRank', 'nomenclaturalStatus', 'taxonomicStatus', 'acceptedNameUsageID']).rename(columns={'taxonID': 'parentNameUsageID', 'scientificName': 'parentScientificName', 'taxonRank': 'parentTaxonRank', 'nomenclaturalStatus': 'parentNomenclaturalStatus', 'taxonomicStatus': 'parentTaxonomicStatus', 'acceptedNameUsageID': 'parentAcceptedNameUsageID'}).set_index(['parentNameUsageID'])
species_df = wfo_df_grouped_by_taxonRank.get_group('species').join(parent_df, on=['parentNameUsageID'], how='left')
display(species_df)

taxonID                     scientificName taxonRank  \
316954   319089                      Aa achalensis   species   
756946   760991                     Aa argyrolepis   species   
918422   922666                      Aa aurantiaca   species   
923805   928062                        Aa calceata   species   
947494   951869                      Aa colombiana   species   
...         ...                                ...       ...   
404334   406838  × Trichoechinopsis × imperialis   species   
404336   406840        × Trichomoza × roseiflora   species   
412580   415123        × Tridentapelia × bijliae   species   
404365   406869      × Turbiniphora × panarottoi   species   
595259   598611             × Vandoglossum moorei   species   

        parentNameUsageID       family             genus nomenclaturalStatus  \
316954         4000000001  Orchidaceae                Aa               Valid   
756946         4000000001  Orchidaceae                Aa               Valid   
918422         4000000001  Orchidaceae                Aa               Valid   
923805         4000000001  Orchidaceae                Aa               Valid   
947494         4000000001  Orchidaceae                Aa               Valid   
...                   ...          ...               ...                 ...   
404334         4000038988    Cactaceae  Trichoechinopsis               Valid   
404336         4000039013    Cactaceae        Trichomoza               Valid   
412580         4001303442  Apocynaceae     Tridentapelia               Valid   
404365         4000039594    Cactaceae      Turbiniphora               Valid   
595259         4000040002  Orchidaceae      Vandoglossum               Valid   

       taxonomicStatus  acceptedNameUsageID parentScientificName  \
316954        Accepted                    0                   Aa   
756946        Accepted                    0                   Aa   
918422        Accepted                    0                   Aa   
923805        Accepted                    0                   Aa   
947494        Accepted                    0                   Aa   
...                ...                  ...                  ...   
404334        Accepted                    0  × Trichoechinopsis   
404336        Accepted                    0        × Trichomoza   
412580        Accepted                    0     × Tridentapelia   
404365        Accepted                    0      × Turbiniphora   
595259        Accepted                    0      × Vandoglossum   

       parentTaxonRank parentNomenclaturalStatus parentTaxonomicStatus  \
316954           genus                     Valid              Accepted   
756946           genus                     Valid              Accepted   
918422           genus                     Valid              Accepted   
923805           genus                     Valid              Accepted   
947494           genus                     Valid              Accepted   
...                ...                       ...                   ...   
404334           genus                     Valid              Accepted   
404336           genus                     Valid              Accepted   
412580           genus                     Valid              Accepted   
404365           genus                     Valid              Accepted   
595259           genus                     Valid              Accepted   

        parentAcceptedNameUsageID  
316954                        0.0  
756946                        0.0  
918422                        0.0  
923805                        0.0  
947494                        0.0  
...                           ...  
404334                        0.0  
404336                        0.0  
412580                        0.0  
404365                        0.0  
595259                        0.0  

[377223 rows x 14 columns]

In [17]:
print(species_df['parentTaxonRank'].value_counts(),'\n')
print(f'We have {len(species_df)} valid and accepted species.')
print(f'and {species_df['parentTaxonRank'].value_counts().sum()} of those have valid and accepted parents.')
# --- 
# save species with parent=(genus, Valid, Accepted) in a separate frame to insert later
# for the rest of the species, check if they have a valid genus listed.
# if yes, add them to the df with the genus listed as parent

parentTaxonRank
genus         372410
section         3331
subsection      1035
subgenus         279
subseries         78
series            71
Name: count, dtype: int64 

We have 377223 valid and accepted species.
and 377204 of those have valid and accepted parents.


In [25]:
# it looks like all species have genus and family listed in the columns genus and family (a.k.a. no empty entry):
print(pd.isnull(species_df['genus']).value_counts())
print(pd.isnull(species_df['family']).value_counts())

genus
False    377223
Name: count, dtype: int64
family
False    377223
Name: count, dtype: int64


In [53]:
# ---- phyla dataframe ----
phyla_df = wfo_df_grouped_by_taxonRank.get_group('phylum').filter(['taxonID', 'scientificName']).merge(phyla_english_names_df, how='left', on='scientificName')
display(phyla_df)

taxonID    scientificName   phylum_englishName
0  9949999999       Angiosperms     Flowering plants
1  4100002134  Anthocerotophyta            Hornworts
2  9949999996         Bryophyta               Mosses
3  4100003333       Cycadophyta          Cycadophyta
4  4100003332       Ginkgophyta               Ginkgo
5  4100003337    Lycopodiophyta             Lycopods
6  4100002135   Marchantiophyta           Liverworts
7  4100003331         Pinophyta  Conifers and allies
8  4100003336    Polypodiophyta                Ferns

In [69]:
# ---- classes dataframe ----
classes_df = wfo_df_grouped_by_taxonRank.get_group('class').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(phyla_df.filter(['taxonID', 'scientificName']), how='left', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_phylum")).filter(['taxonID', 'scientificName', 'taxonID_parent_phylum', 'scientificName_parent_phylum'])
display(classes_df)

taxonID        scientificName  taxonID_parent_phylum  \
0   9850000001     Andreaeobryopsida             9949999996   
1   9850000002         Andreaeopsida             9949999996   
2   4100000009      Anthocerotopsida             4100002134   
3   4100002058             Bryopsida             9949999996   
4   4100002548           Cycadopsida             4100003333   
5   4100002550          Ginkgoopsida             4100003332   
6   4100000011      Haplomitriopsida             4100002135   
7   4100000012     Jungermanniopsida             4100002135   
8   4100000010  Leiosporocerotopsida             4100002134   
9   4100003338        Lycopodiopsida             4100003337   
10  4100000013       Marchantiopsida             4100002135   
11  4100002552             Pinopsida             4100003331   
12  4100003341        Polypodiopsida             4100003336   
13  9850000004       Polytrichopsida             9949999996   
14  9850000005          Sphagnopsida             9949999996   
15  9850000006          Takakiopsida             9949999996   

   scientificName_parent_phylum  
0                     Bryophyta  
1                     Bryophyta  
2              Anthocerotophyta  
3                     Bryophyta  
4                   Cycadophyta  
5                   Ginkgophyta  
6               Marchantiophyta  
7               Marchantiophyta  
8              Anthocerotophyta  
9                Lycopodiophyta  
10              Marchantiophyta  
11                    Pinophyta  
12               Polypodiophyta  
13                    Bryophyta  
14                    Bryophyta  
15                    Bryophyta

In [70]:
# ---- subclasses dataframe ----
subclasses_df = wfo_df_grouped_by_taxonRank.get_group('subclass').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(classes_df.filter(['taxonID', 'scientificName']), how='left', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_class")).filter(['taxonID', 'scientificName', 'taxonID_parent_class', 'scientificName_parent_class'])
display(subclasses_df)

taxonID   scientificName  taxonID_parent_class  \
0   4100002070   Anthocerotidae            4100000009   
1   4100002073        Blasiidae            4100000013   
2   9800000001          Bryidae            4100002058   
3   9800000002     Buxbaumiidae            4100002058   
4   4100002553      Cupressidae            4100002552   
5   4100002549        Cycadidae            4100002548   
6   1000044985  Dendrocerotidae            4100000009   
7   9800000006       Dicranidae            4100002058   
8   9800000003     Diphysciidae            4100002058   
9   4100003340      Equisetidae            4100003341   
10  9800000004       Funariidae            4100002058   
11  9800000007    Gigaspermidae            4100002058   
12  4100002551       Ginkgoidae            4100002550   
13  4100002555         Gnetidae            4100002552   
14  4100002071   Haplomitriidae            4100000011   
15  4100002076  Jungermanniidae            4100000012   
16  4100003343      Marattiidae            4100003341   
17  4100002074    Marchantiidae            4100000013   
18  4100002077     Metzgeriidae            4100000012   
19  1000044986   Notothyladidae            4100000009   
20  4100003342   Ophioglossidae            4100003341   
21  4100002078        Pelliidae            4100000012   
22  4100002554          Pinidae            4100002552   
23  4100003344     Polypodiidae            4100003341   
24  9800000005        Timmiidae            4100002058   
25  4100002072       Treubiidae            4100000011   

   scientificName_parent_class  
0             Anthocerotopsida  
1              Marchantiopsida  
2                    Bryopsida  
3                    Bryopsida  
4                    Pinopsida  
5                  Cycadopsida  
6             Anthocerotopsida  
7                    Bryopsida  
8                    Bryopsida  
9               Polypodiopsida  
10                   Bryopsida  
11                   Bryopsida  
12                Ginkgoopsida  
13                   Pinopsida  
14            Haplomitriopsida  
15           Jungermanniopsida  
16              Polypodiopsida  
17             Marchantiopsida  
18           Jungermanniopsida  
19            Anthocerotopsida  
20              Polypodiopsida  
21           Jungermanniopsida  
22                   Pinopsida  
23              Polypodiopsida  
24                   Bryopsida  
25            Haplomitriopsida

In [71]:
# ---- superorders dataframe ----
superorders_df = wfo_df_grouped_by_taxonRank.get_group('superorder').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(subclasses_df.filter(['taxonID', 'scientificName']), how='left', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_subclass")).filter(['taxonID', 'scientificName', 'taxonID_parent_subclass', 'scientificName_parent_subclass'])
display(superorders_df)

taxonID scientificName  taxonID_parent_subclass  \
0  9500000001        Bryanae               9800000001   
1  9500000002       Hypnanae               9800000001   

  scientificName_parent_subclass  
0                        Bryidae  
1                        Bryidae

In [91]:
# ---- orders dataframe ----
order_parents = {'phylum': phyla_df, 'class': classes_df, 'subclass': subclasses_df, 'superorder': superorders_df}
# ---- 4 orders_dfs, one for each parent category
orders_dfs = [wfo_df_grouped_by_taxonRank.get_group('order').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(order_parents[key].filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, f"_parent_{key}")) for key in order_parents]
all_columns = set().union(*[df.columns for df in orders_dfs])

# display(orders_dfs)
orders_df = pd.concat([df.filter(['taxonID', 'scientificName', 'parentNameUsageID']) for df in orders_dfs], axis=0)
display(orders_df)

taxonID  scientificName  parentNameUsageID
0   9000000003        Acorales         9949999999
1   9000000013     Alismatales         9949999999
2   9000000014    Amborellales         9949999999
3   9000000022         Apiales         9949999999
4   9000000023    Aquifoliales         9949999999
..         ...             ...                ...
9   4100000612  Orthodontiales         9500000001
10  9000000366  Orthotrichales         9500000001
11  9000000439   Ptychomniales         9500000002
12  9000000457   Rhizogoniales         9500000001
13  9000000502     Splachnales         9500000001

[142 rows x 3 columns]

In [93]:
# ---- families dataframe ----
families_df = wfo_df_grouped_by_taxonRank.get_group('family').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(orders_df.filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_order")).filter(['taxonID', 'scientificName', 'taxonID_parent_order', 'scientificName_parent_order'])
display(families_df)

taxonID   scientificName  taxonID_parent_order  \
0    7000000001      Acanthaceae            9000000283   
1    7000000002      Achariaceae            9000000311   
2    7000000003  Achatocarpaceae            9000000088   
3    7000000004        Acoraceae            9000000003   
4    7000000005    Acrobolbaceae            9000000278   
..          ...              ...                   ...   
712  7000000649       Xyridaceae            9000000415   
713  7000000650        Zamiaceae            9000000149   
714  7000000651    Zingiberaceae            9000000577   
715  7000000652      Zosteraceae            9000000013   
716  7000000653   Zygophyllaceae            9000000581   

    scientificName_parent_order  
0                      Lamiales  
1                  Malpighiales  
2                Caryophyllales  
3                      Acorales  
4               Jungermanniales  
..                          ...  
712                      Poales  
713                   Cycadales  
714                Zingiberales  
715                 Alismatales  
716               Zygophyllales  

[717 rows x 4 columns]

In [94]:
# ---- subfamilies dataframe ----
subfamilies_df = wfo_df_grouped_by_taxonRank.get_group('subfamily').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(families_df.filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_family")).filter(['taxonID', 'scientificName', 'taxonID_parent_family', 'scientificName_parent_family'])
display(subfamilies_df)

taxonID     scientificName  taxonID_parent_family  \
0   6500000985    Acrosanthoideae             7000000010   
1   6500000986         Aizooideae             7000000010   
2   6500000984       Alpinioideae             7000000651   
3   1000002334        Arbutoideae             7000000218   
4   4100002170     Asphodeloideae             7000000659   
..         ...                ...                    ...   
85  6500000529    Woodwardioideae             7000000077   
86  4100000025  Wunderlichioideae             7000000146   
87  1000000028  Xanthoceratoideae             7000000544   
88  6500000775  Xanthorrhoeoideae             7000000659   
89  6500000983     Zingiberoideae             7000000651   

   scientificName_parent_family  
0                     Aizoaceae  
1                     Aizoaceae  
2                 Zingiberaceae  
3                     Ericaceae  
4                 Asphodelaceae  
..                          ...  
85                  Blechnaceae  
86                   Asteraceae  
87                  Sapindaceae  
88                Asphodelaceae  
89                Zingiberaceae  

[90 rows x 4 columns]

In [95]:
# ---- tribes dataframe ----
tribes_df = wfo_df_grouped_by_taxonRank.get_group('tribe').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(subfamilies_df.filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_subfamily")).filter(['taxonID', 'scientificName', 'taxonID_parent_subfamily', 'scientificName_parent_subfamily'])
display(tribes_df)

taxonID   scientificName  taxonID_parent_subfamily  \
0   4100002171           Aloeae                4100002170   
1   5000001280        Alpinieae                6500000984   
2   5000001283  Anisostigmateae                6500000990   
3   4100000064      Anthemideae                4100000023   
4   4100002037   Anthocercideae                4100002036   
..         ...              ...                       ...   
79  5000000588        Tamijieae                6500000617   
80  4100000084   Tarchonantheae                4100000027   
81  4000040168       Vernonieae                4100000036   
82  5000000850   Wunderlichieae                4100000025   
83  5000001282      Zingibereae                6500000983   

   scientificName_parent_subfamily  
0                   Asphodeloideae  
1                     Alpinioideae  
2                     Sesuvioideae  
3                      Asteroideae  
4                   Nicotianoideae  
..                             ...  
79                    Tamijioideae  
80                     Carduoideae  
81                   Vernonioideae  
82               Wunderlichioideae  
83                  Zingiberoideae  

[84 rows x 4 columns]

In [96]:
# ---- subtribes dataframe ----
subtribes_df = wfo_df_grouped_by_taxonRank.get_group('subtribe').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(tribes_df.filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_tribe")).filter(['taxonID', 'scientificName', 'taxonID_parent_tribe', 'scientificName_parent_tribe'])
display(subtribes_df)

taxonID    scientificName  taxonID_parent_tribe  \
0    4100001320  Adenostemmatinae            4100000042   
1    4100002022     Afroasterinae            4100000054   
2    4500000250        Ageratinae            4100000042   
3    4500000251         Alomiinae            4100000042   
4    4100001352       Ambrosiinae            5000000795   
..          ...               ...                   ...   
165  4100001275       Vernoniinae            4000040168   
166  4500000394       Warioniinae            4100000044   
167  4100001932     Xerantheminae            4100000062   
168  4500000047      Zaluzaniinae            5000000795   
169  4500000230         Zinniinae            5000000795   

    scientificName_parent_tribe  
0                   Eupatorieae  
1                      Astereae  
2                   Eupatorieae  
3                   Eupatorieae  
4                   Heliantheae  
..                          ...  
165                  Vernonieae  
166                  Cichorieae  
167                    Cardueae  
168                 Heliantheae  
169                 Heliantheae  

[170 rows x 4 columns]

In [106]:
# ---- genera dataframe ---- columns = ['taxonID', 'scientificName', 'family', 'familyID']
genus_parents = {'phylum': phyla_df, 'class': classes_df, 'subclass': subclasses_df, 'superorder': superorders_df, 'order': orders_df, 'family': families_df, 'subfamily': subfamilies_df, 'tribe': tribes_df, 'subtribe': subtribes_df}
# ---- 9 orders_dfs, one for each parent category
genera_dfs = [wfo_df_grouped_by_taxonRank.get_group('genus').filter(['taxonID', 'scientificName', 'family', 'genus', 'parentNameUsageID']).merge(genus_parents[key].filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_genus")) for key in genus_parents]
# display(genera_dfs)
genera_df = pd.concat([df.filter(['taxonID', 'scientificName', 'parentNameUsageID']) for df in genera_dfs], axis=0)
display(genera_df)

taxonID       scientificName  parentNameUsageID
0    4000000001                   Aa         7000000429
1    4000046264                Aakia         7000000483
2    4000000016               Abatia         7000000540
3    4000000024                Abdra         7000000082
4    4000000030        Abeliophyllum         7000000422
..          ...                  ...                ...
852  4000041233               Zinnia         4500000230
853  4000041254               Zoegea         4100001338
854  4000041357            Zyrphelis         4100001316
855  4000041358              Zyzyxia         4100001302
856  4000006703  × Carduogalactites         4100001278

[17910 rows x 3 columns]

In [61]:
print(genera_df.loc[genera_df['scientificName'] == 'Athyrium'])

            taxonID scientificName        family    familyID
1408638  4000003575       Athyrium  Aspleniaceae  7000000051
1408637  4000003569       Athyrium   Athyriaceae  7000000055


In [107]:
# ---- subgenus dataframe ----
subgenera_df = wfo_df_grouped_by_taxonRank.get_group('subgenus').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(genera_df.filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_genus")).filter(['taxonID', 'scientificName', 'taxonID_parent_genus', 'scientificName_parent_genus'])
display(subgenera_df)

taxonID                    scientificName  taxonID_parent_genus  \
0   3500004052             Aizoon subgen. Aizoon            4000001055   
1   3500004053           Aizoon subgen. Capensia            4000001055   
2   3500004054            Aizoon subgen. Galenia            4000001055   
3   3500004056           Aizoon subgen. Karooica            4000001055   
4   3500004057           Aizoon subgen. Kolleria            4000001055   
..         ...                               ...                   ...   
66  3500004144   Tetragonia subgen. Tetragonella            4000037907   
67  3500004145     Tetragonia subgen. Tetragonia            4000037907   
68  3500004148  Tetragonia subgen. Tetragonoides            4000037907   
69  3500004150      Trianthema subgen. Papularia            4000038870   
70  3500004152     Trianthema subgen. Trianthema            4000038870   

   scientificName_parent_genus  
0                       Aizoon  
1                       Aizoon  
2                       Aizoon  
3                       Aizoon  
4                       Aizoon  
..                         ...  
66                  Tetragonia  
67                  Tetragonia  
68                  Tetragonia  
69                  Trianthema  
70                  Trianthema  

[71 rows x 4 columns]

In [176]:
# ---- sections dataframe ---- 
section_parents = {'genus': genera_df, 'subgenus': subgenera_df}
# ---- 2 section_dfs, one for each parent category
sections_dfs = [wfo_df_grouped_by_taxonRank.get_group('section').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(section_parents[key].filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, f"_parent_{key}")) for key in section_parents]
# display(sections_df)
sections_df = pd.concat([df.filter(['taxonID', 'scientificName', 'parentNameUsageID']) for df in sections_dfs], axis=0)
display(sections_df)

taxonID                scientificName  parentNameUsageID
0   1000041355         Acmella sect. Acmella         4000000312
1   4100000113   Anisoptera sect. Anisoptera         4000002200
2   1000040263      Anisoptera sect. Glabrae         4000002200
3   1000039816         Begonia sect. Alicida         4000004308
4   1000039471   Begonia sect. Apterobegonia         4000004308
..         ...                           ...                ...
57  3400004945  Rhododendron sect. Tsutsutsi         1000040525
58  3400012108  Riedelia sect. Coralliophyta         3500004009
59  3400012109        Riedelia sect. Cornuta         3500004009
60  3400012110    Riedelia sect. Geocharides         3500004009
61  3400012111      Riedelia sect. Macrantha         3500004009

[267 rows x 3 columns]

In [123]:
# ---- subsections dataframe ----
subsection_parents = {'genus': genera_df, 'section': sections_df}
# ---- 2 subsection_dfs, one for each parent category
subsections_dfs = [wfo_df_grouped_by_taxonRank.get_group('subsection').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(subsection_parents[key].filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, f"_parent_{key}")) for key in subsection_parents]
# display(subsections_dfs)
subsections_df = pd.concat([df.filter(['taxonID', 'scientificName', 'parentNameUsageID']) for df in subsections_dfs], axis=0)
display(subsections_df)

taxonID                     scientificName  parentNameUsageID
0   1000029004     Solidago subsect. Brachychaeta         4000035732
1   1000029005     Solidago subsect. Drummondiani         4000035732
2   3000000102  Solidago subsect. Glomeruliflorae         4000035732
3   3000001047          Solidago subsect. Humiles         4000035732
4   3000002927          Solidago subsect. Junceae         4000035732
..         ...                                ...                ...
68  3000003222    Riedelia subsect. Subulocalyces         3400012109
69  1000050961   Rubroshorea subsect. Auriculatae         1000056268
70  1000056269   Rubroshorea subsect. Rubroshorea         1000056268
71  1000000059             Shorea subsect. Mutica         1000000057
72  1000000066             Shorea subsect. Shorea         1000000065

[86 rows x 3 columns]

In [126]:
# ---- series dataframe ----
series_parents = {'genus': genera_df, 'section': sections_df}
# ---- 2 series_df, one for each parent category
series_dfs = [wfo_df_grouped_by_taxonRank.get_group('series').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(series_parents[key].filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, f"_parent_{key}")) for key in series_parents]
# display(subsections_dfs)
series_df = pd.concat([df.filter(['taxonID', 'scientificName', 'parentNameUsageID']) for df in series_dfs], axis=0)
display(series_df)

taxonID                   scientificName  parentNameUsageID
0   1000002456    Gaultheria ser. Trichophyllae         4000015402
1   1000041375          Senecio ser. Nemorenses         4000035060
2   2800002943             Senna ser. Aculeatae         4000035076
3   2800003003              Senna ser. Aphyllae         4000035076
4   2800003659            Solidago ser. Argutae         4000035732
5   2800004439      Solidago ser. Brachychaetae         4000035732
6   1000029000         Solidago ser. Canadenses         4000035732
7   1000024771       Solidago ser. Macrophyllae         4000035732
8   2800004440        Solidago ser. Spectabiles         4000035732
9   1000029002        Solidago ser. Tortifoliae         4000035732
10  1000029003         Solidago ser. Ulmifoliae         4000035732
11  2800004711            Solidago ser. Venosae         4000035732
0   2800002381       Baccharis ser. Canescentes         3400006074
1   2800002401      Baccharis ser. Colliculatae         3400000721
2   2800002378      Baccharis ser. Curitybenses         3400000721
3   2800002380     Baccharis ser. Gnaphalioides         3400006074
4   2800002379  Baccharis ser. Tarchonanthoides         3400000721
5   1000029001          Solidago ser. Puberulae         1000036096
6   2800000099              Styrax ser. Benzoin         3400002933
7   2800004458                Styrax ser. Cyrta         1000000099
8   1000000100               Styrax ser. Styrax         1000000099
9   2800000358             Styrax ser. Valvatae         3400002933

In [127]:
# ---- subseries dataframe ----
subseries_df = wfo_df_grouped_by_taxonRank.get_group('subseries').filter(['taxonID', 'scientificName', 'parentNameUsageID']).merge(series_df.filter(['taxonID', 'scientificName']), how='inner', left_on='parentNameUsageID', right_on='taxonID', suffixes=(None, "_parent_series")).filter(['taxonID', 'scientificName', 'taxonID_parent_series', 'scientificName_parent_series'])
display(subseries_df)

taxonID             scientificName  taxonID_parent_series  \
0  2700000233  Styrax subser. Foveolaria             2800000358   
1  2700000232    Styrax subser. Latifoli             2800000358   

  scientificName_parent_series  
0         Styrax ser. Valvatae  
1         Styrax ser. Valvatae

In [178]:
# ---- species dataframe ---- columns = ['taxonID', 'scientificName', 'genus', 'family', 'genusID']
genera_taxonID = genera_df.filter(['taxonID', 'scientificName']).rename(columns={'taxonID': 'parentNameUsageID', 'scientificName': 'genus'}).set_index(['parentNameUsageID', 'genus'])
species_df = wfo_df_grouped_by_taxonRank.get_group('species').filter(['taxonID', 'scientificName', 'parentNameUsageID', 'genus', 'family']).join(genera_taxonID, on=['parentNameUsageID', 'genus'], how='inner')
#.join(genera_taxonID, on='genus', how='inner')
# genera_taxonID2 = genera_df.filter(['taxonID', 'scientificName']).rename(columns={'taxonID': 'parentNameUsageID', 'scientificName': 'genus'}).set_index(['parentNameUsageID', 'genus'])
#display(species_df.join(genera_taxonID2, on=['parentNameUsageID', 'genus'], how='inner'))
#display(species_df.join(genera_taxonID, on='genus', how='inner'))

display(species_df)

taxonID         scientificName  parentNameUsageID      genus  \
316954    319089          Aa achalensis         4000000001         Aa   
756946    760991         Aa argyrolepis         4000000001         Aa   
918422    922666          Aa aurantiaca         4000000001         Aa   
923805    928062            Aa calceata         4000000001         Aa   
947494    951869          Aa colombiana         4000000001         Aa   
...          ...                    ...                ...        ...   
26751      26166      Zyrphelis montana         4000041357  Zyrphelis   
81214      80864  Zyrphelis perezioides         4000041357  Zyrphelis   
92607      92281    Zyrphelis taxifolia         4000041357  Zyrphelis   
1312510  1334611         Zyzyura mayana         4001303479    Zyzyura   
139119    138947      Zyzyxia lundellii         4000041358    Zyzyxia   

              family  
316954   Orchidaceae  
756946   Orchidaceae  
918422   Orchidaceae  
923805   Orchidaceae  
947494   Orchidaceae  
...              ...  
26751     Asteraceae  
81214     Asteraceae  
92607     Asteraceae  
1312510   Asteraceae  
139119    Asteraceae  

[372205 rows x 5 columns]

### 2. Inserting data from df into postgres

In [168]:
db_params = {
    'dbname': 'WhatPlant_DB',
    'user': 'postgres',
    'password': 'codersbay',
    'host': 'localhost',
    'port': '5432'
}

conn = psycopg2.connect(
        dbname=db_params['dbname'],
        user=db_params['user'],
        password=db_params['password'],
        host=db_params['host'],
        port=db_params['port']
    )
cur = conn.cursor()

In [169]:
insert_into_phyla_query = '''   INSERT INTO WhatPlant.phyla (phylum_id, phylum_name, phylum_english_name) 
                                VALUES (%s, %s, %s); '''

insert_into_classes_query = ''' INSERT INTO WhatPlant.classes (class_id, class_name, phylum_id) 
                                VALUES (%s, %s, %s); '''

insert_into_subclasses_query = ''' INSERT INTO WhatPlant.subclasses (subclass_id, subclass_name, class_id) 
                                   VALUES (%s, %s, %s); '''

insert_into_orders_query = '''   INSERT INTO WhatPlant.orders (order_id, order_name) 
                                            VALUES (%s, %s); '''
insert_into_orders_query_w_subclass = '''   INSERT INTO WhatPlant.orders (order_id, order_name, subclass_id) 
                                            VALUES (%s, %s, %s); '''
insert_into_orders_query_w_class = '''  INSERT INTO WhatPlant.orders (order_id, order_name, class_id) 
                                        VALUES (%s, %s, %s); '''
insert_into_orders_query_w_phylum = '''     INSERT INTO WhatPlant.orders (order_id, order_name, phylum_id) 
                                            VALUES (%s, %s, %s); '''

insert_into_families_query = '''  INSERT INTO WhatPlant.families (family_id, family_name, order_id)
                                VALUES (%s, %s, %s); '''

insert_into_genera_query = '''  INSERT INTO WhatPlant.genera (genus_id, genus_name, parent_id)
                                VALUES (%s, %s, %s); '''

insert_into_species_query = ''' INSERT INTO WhatPlant.species (species_id, species_name, genus_name, family_name)
                                    VALUES (%s, %s, %s, %s); '''

insert_into_species_synonyms_query = '''    INSERT INTO WhatPlant.species_synonyms 
                                            (species_synonym_id, species_synonym_name, synonym_for)
                                            VALUES (%s, %s, %s); '''

In [170]:
for _, row in phyla_df.iterrows():
    phylum_data = (row['taxonID'], row['scientificName'], row['phylum_englishName'])
    cur.execute(insert_into_phyla_query, phylum_data)

for _, row in classes_df.iterrows():
    class_data = (row['taxonID'], row['scientificName'], row['taxonID_parent_phylum'])
    cur.execute(insert_into_classes_query, class_data)

for _, row in subclasses_df.iterrows():
    subclass_data = (row['taxonID'], row['scientificName'], row['taxonID_parent_class'])
    cur.execute(insert_into_subclasses_query, subclass_data)

In [171]:
for _, row in orders_df.iterrows():
    # order_data = (row['taxonID'], row['scientificName'], row['parentNameUsageID'])
    order_data = (row['taxonID'], row['scientificName'])
    cur.execute(insert_into_orders_query, order_data)

In [172]:
for _, row in families_df.iterrows():
    family_data = (row['taxonID'], row['scientificName'], row['taxonID_parent_order'])
    cur.execute(insert_into_families_query, family_data)

In [173]:
for _, row in genera_df.iterrows():
    genus_data = (row['taxonID'], row['scientificName'], row['parentNameUsageID'])
    cur.execute(insert_into_genera_query, genus_data)

In [174]:
for _, row in species_df.iterrows():
    species_data = (row['taxonID'], row['scientificName'], row['genus'], row['family'])
    cur.execute(insert_into_species_query, species_data)

In [175]:
conn.commit()
cur.close()
conn.close()